pkg/clioptions/clusterdiscovery/cluster.go

```diff
@@ -56,6 +56,8 @@ type ClusterConfiguration struct {
 	// and dual-stack-specific tests are run
 	HasIPv4 bool
 	HasIPv6 bool
+	// IPFamily defines default IP stack of the cluster, replaces upstream getDefaultClusterIPFamily
+	IPFamily string
```

Comment:
Seems like it should be `PrimaryIPFamily`?

Also, `HasIPv4`+`HasIPv6`+`IPFamily` is redundant (and allows impossible configurations like `hasIPv6: false, ipFamily: "ipv6"`). We don't want to get rid of the old fields, for compatibility, but it would probably be nicer to make this `PrimaryIPFamily string` (or `PrimaryIPFamily v1.Family`?) plus `DualStack bool` and then deprecate `HasIPv4` and `HasIPv6`?

Reply:
`IPFamily` matches `context.IPFamily`  Changing the name would confuse.

The only use case of `context.IPFamily` in e2e seems to be`ClusterIsIPv6` which is used frequently.

```
// ClusterIsIPv6 returns true if the cluster is IPv6
func (tc TestContextType) ClusterIsIPv6() bool {
	return tc.IPFamily == "ipv6"
}
```

So, assume IPv4 unless IPFamily is IPv6.  Then to detect dualstack something else.

```
	if c.IPFamily == "ipv6" {
		skips = append(skips, "[Feature:Networking-IPv4]")
	} else {
		skips = append(skips, "[Feature:Networking-IPv6]")
	}
	if !c.HasIPv4 || !c.HasIPv6 {
		// lack of "]" is intentional; this matches multiple tags
		skips = append(skips, "[Feature:IPv6DualStack")
	
```

with consts.

Reply:
err, well 

```go
	if !c.DualStack {
		// lack of "]" is intentional; this matches multiple tags
		skips = append(skips, "[Feature:IPv6DualStack")
		if c.IPFamily == strings.ToLower(string(corev1.IPv6Protocol)) {
			skips = append(skips, "[Feature:Networking-IPv4]")
		} else {
			skips = append(skips, "[Feature:Networking-IPv6]")
		}
	}
	// else DualStack, run all the tests.
```

Reply:
> 
> Also, `HasIPv4`+`HasIPv6`+`IPFamily` is redundant (and allows impossible configurations like `hasIPv6: false, ipFamily: "ipv6"`). 

I'm not sure what to do in the case of no discovery.  The `tc.Context` has an IPFamily but we seem to want to skip `ipv4` tests if there is no discovery, which implies that `HasIPv4` can be false while `IPFamily == "ipv4"`


```go
		{
			name:                 "complex override without discovery",
			provider:             `{"type":"aws","region":"us-east-2","zone":"us-east-2a","multimaster":false,"multizone":true}`,
			discoveredPlatform:   nil,
			optionalCapabilities: configv1.KnownClusterVersionCapabilities,
			expectedConfig:       `{"type":"aws","ProjectID":"","Region":"us-east-2","Zone":"us-east-2a","NumNodes":0,"MultiMaster":false,"MultiZone":true,"Zones":null,"ConfigFile":"","Disconnected":false,"SingleReplicaTopology":false,"NetworkPlugin":"","HasIPv4":false,"HasIPv6":false,"HasSCTP":false,"IsProxied":false,"IsIBMROKS":false,"HasNoOptionalCapabilities":false}`,
			runTests:             sets.NewString("everyone", "not-gce", "not-sdn", "not-multitenant", "online", "requires-optional-cap"),
		},

```